README.md
# Swift Mailer AWS Simple Email Service (SES) Transport

Allows for sending email using [Swift Mailer](https://swiftmailer.symfony.com/) over HTTP uses [AWS Simple Email Service](https://aws.amazon.com/ses/)

```php
// Create the SES client
$ses_client = SESTransportFactory::buildSESClient(
    [
        'credentials' => [
            'key'    => 'AWS_ACCESS_KEY',
            'secret' => 'AWS_SECRET',
        ],
    ]
);

// Create the transport
$transport = SESTransportFactory::buildSESTransport(
    $ses_client,
    new OperationTimer(new NullMetricsAgent()),
    new NullLogger()
);

// Create the mailer
$mailer = new Swift_Mailer($transport);
```


CHANGELOG.md

### Unreleased

## v1.1.0 (2022-10-16)

* Support PHP 8.1

## v1.0.2 (2021-12-14)

* Increase entropy of unique filename so 2 emails processed in quick succession do not overwrite each other

## v1.0.1 (2021-04-21)

* Support PHP8

## v1.0.0 (2021-03-20)

* AWS SES SwiftMailer transport release


test/mock/SES/Transport/SpyingSESClient.php
<?php
/**
 * @author    Craig Gosman <craig@ingenerator.com>
 * @licence   proprietary
 */


namespace test\mock\Ingenerator\SwiftMailer\SES\Transport;


use Aws\Result;
use Aws\Ses\SesClient;

class SpyingSESClient extends SesClient
{
    protected array $sent_mails = [];
    protected bool $should_throw = FALSE;

    public function __construct()
    {
    }

    public static function sendRawEmailWillThrow()
    {
        $s               = new self;
        $s->should_throw = TRUE;

        return $s;
    }

    public function getSentMails(): array
    {
        return $this->sent_mails;
    }

    public function sendRawEmail(array $args = []): Result
    {
        if ($this->should_throw) {
            throw new \Exception('You asked me to throw');
        }

        $this->sent_mails[] = $args;

        return new Result();
    }

}


test/unit/SES/Transport/SESTransportTest.php
<?php
/**
 * @author    Craig Gosman <craig@ingenerator.com>
 * @licence   proprietary
 */

namespace test\unit\Ingenerator\SwiftMailer\SES\Transport;

use Ingenerator\PHPUtils\Monitoring\ArrayMetricsAgent;
use Ingenerator\PHPUtils\Monitoring\OperationTimer;
use Ingenerator\SwiftMailer\Common\Transport\EmailDeliveryFailedException;
use Ingenerator\SwiftMailer\SES\Transport\SESTransport;
use InvalidArgumentException;
use PHPUnit\Framework\Assert;
use PHPUnit\Framework\MockObject\MockObject;
use PHPUnit\Framework\TestCase;
use Psr\Log\LogLevel;
use Psr\Log\Test\TestLogger;
use Swift_Events_EventDispatcher;
use Swift_Events_SendEvent;
use Swift_Events_SendListener;
use Swift_Events_SimpleEventDispatcher;
use Swift_Message;
use Swift_Mime_SimpleMessage;
use test\mock\Ingenerator\SwiftMailer\SES\Transport\SpyingSESClient;

class DummySwiftPlugin implements Swift_Events_SendListener
{
    protected array $got_events;

    public function assertGotEvents(array $expect)
    {
        Assert::assertSame($expect, $this->got_events);
    }

    public function beforeSendPerformed(Swift_Events_SendEvent $evt)
    {
        $this->got_events[] = [__FUNCTION__, $evt];
    }

    public function sendPerformed(Swift_Events_SendEvent $evt)
    {
        $this->got_events[] = [__FUNCTION__, $evt];
    }

}

class DummySwift_Message extends Swift_Message
{
    private string $dummy_mime_string;

    public static function withDummyMimeString(string $string): DummySwift_Message
    {
        $s                    = new self();
        $s->dummy_mime_string = $string;

        return $s;
    }

    public function toString(): string
    {
        return $this->dummy_mime_string;
    }
}

class SESTransportTest extends TestCase
{
    protected Swift_Events_EventDispatcher $event_dispatcher;

    protected SpyingSESClient $ses_client;

    protected ArrayMetricsAgent $metrics;

    protected TestLogger $log;

    public function test_it_is_initialisable()
    {
        $this->assertInstanceOf(SESTransport::class, $this->newSubject());
    }

    public function test_it_is_started()
    {
        $this->assertTrue($this->newSubject()->isStarted());
    }

    public function test_it_throws_with_more_than_one_from_address()
    {
        $message = new Swift_Message;
        $message->addFrom('foo@bar.com');
        $message->addFrom('other@bar.com');
        $subject = $this->newSubject();
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('multiple `from` address');
        $subject->send($message);
    }

    public function test_it_throws_with_more_than_one_to_address()
    {
        $message = new Swift_Message;
        $message->addTo('foo@bar.com');
        $message->addTo('other@bar.com');
        $subject = $this->newSubject();
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('multiple `to` address');
        $subject->send($message);
    }

    public function test_its_register_plugin_just_binds_to_event_listener()
    {
        $plugin  = new DummySwiftPlugin;
        $subject = $this->newSubject();
        $subject->registerPlugin($plugin);

        $evt = $this->event_dispatcher->createSendEvent($subject, $this->getDummy(Swift_Mime_SimpleMessage::class));
        $this->event_dispatcher->dispatchEvent($evt, 'sendPerformed');
        $plugin->assertGotEvents([['sendPerformed', $evt]]);
    }

    public function test_its_send_method_sends()
    {
        $subject = $this->newSubject();
        $message = $this->generateValidMessage();
        $this->assertSame(1, $subject->send($message), 'Number of messages sent does not equal 1');
        $this->assertSame([['RawMessage' => ['Data' => $message->toString()]]], $this->ses_client->getSentMails());
    }

    public function test_its_start_is_no_op()
    {
        $this->assertTrue($this->newSubject()->start());
    }

    public function test_its_stop_is_no_op()
    {
        $this->assertTrue($this->newSubject()->stop());
    }

    public function test_it_emits_timer_metric()
    {
        $subject = $this->newSubject();
        $subject->send($this->generateValidMessage());
        $this->metrics->assertCapturedOneTimer('ses-email-sent', 'ok');
    }

    public function test_it_emits_timer_metric_with_err_src_on_failure()
    {
        $this->ses_client = SpyingSESClient::sendRawEmailWillThrow();
        $subject          = $this->newSubject();
        try {
            $subject->send($this->generateValidMessage());
        } catch (\Exception $e) {
            $this->assertEquals('Email send failed: [Exception] You asked me to throw', $e->getMessage());
        }
        $this->metrics->assertCapturedOneTimer('ses-email-sent', 'err');
    }

    public function test_it_logs_success()
    {
        $subject = $this->newSubject();
        $subject->send($this->generateValidMessage());
        $this->assertSame(
            [
                [
                    'level'   => LogLevel::INFO,
                    'message' => 'Emailed foo@bar.com ()',
                    'context' => [
                        'aws_msg_id' => NULL,
                    ],
                ],
            ],
            $this->log->records
        );
    }

    public function test_it_logs_failure()
    {
        $this->ses_client = SpyingSESClient::sendRawEmailWillThrow();
        $subject          = $this->newSubject();
        try {
            $subject->send($this->generateValidMessage());
        } catch (\Exception $e) {
            $this->assertInstanceOf(EmailDeliveryFailedException::class, $e);
            $this->assertEquals('Email send failed: [Exception] You asked me to throw', $e->getMessage());
            $this->assertSame('You asked me to throw', $e->getPrevious()->getMessage());
        }
        $this->assertSame(
            LogLevel::ERROR,
            $this->log->records[0]['level']
        );
        $this->assertSame(
            'Failed to email `foo@bar.com` - [Exception] You asked me to throw',
            $this->log->records[0]['message']
        );
    }

    protected function generateValidMessage(): DummySwift_Message
    {
        $message = DummySwift_Message::withDummyMimeString('DUMMY MESSAGE');
        $message->addTo('foo@bar.com');
        $message->addFrom('other@bar.com');

        return $message;
    }

    protected function getDummy(string $class): MockObject
    {
        return $this->getMockBuilder($class)
            ->disableOriginalConstructor()
            ->disableOriginalClone()
            ->disableProxyingToOriginalMethods()
            ->getMock();
    }

    protected function newSubject(): SESTransport
    {
        return new SESTransport(
            $this->event_dispatcher,
            $this->ses_client,
            new OperationTimer($this->metrics),
            $this->log
        );
    }

    protected function setUp(): void
    {
        parent::setUp();
        $this->event_dispatcher = new Swift_Events_SimpleEventDispatcher;
        $this->ses_client       = new SpyingSESClient();
        $this->metrics          = new ArrayMetricsAgent();
        $this->log              = new TestLogger();
    }

}


test/unit/SES/DevSupport/SESSimulatorEmailParserTest.php
<?php
/**
 * @author    Craig Gosman <craig@ingenerator.com>
 * @licence   proprietary
 */

namespace test\unit\Ingenerator\SwiftMailer\SES\DevSupport;

use BadMethodCallException;
use Ingenerator\Mailhook\Email;
use Ingenerator\SwiftMailer\SES\DevSupport\SESSimulatorEmailParser;
use PHPUnit\Framework\TestCase;
use Swift_Message;
use UnexpectedValueException;

class SESSimulatorEmailParserTest extends TestCase
{

    public function test_it_is_initialisable()
    {
        $this->assertInstanceOf(SESSimulatorEmailParser::class, $this->newSubject());
    }

    public function test_its_parse_method_throws()
    {
        $subject = $this->newSubject();
        $this->expectException(BadMethodCallException::class);
        $subject->parse('');
    }

    public function test_its_parseSimulatorCapture_throws_if_captured_action_is_not_SendRawEmail()
    {
        $subject = $this->newSubject();
        $this->expectException(UnexpectedValueException::class);
        $this->expectExceptionMessage('Expected AWS action SendRawEmail, got UnexpectedAction');
        $subject->parseSimulatorCapture(['post' => ['Action' => 'UnexpectedAction']]);
    }

    public function test_it_can_parseSimulatorCapture()
    {
        $subject = $this->newSubject();
        $email   = $subject->parseSimulatorCapture(
            [
                'post'     => ['Action' => 'SendRawEmail'],
                'raw_data' => <<<MSG
                                Date: Thu, 30 Jul 2020 22:55:04 +0100
                                Subject: inGenerator - Account Activation
                                From: NoOne <noreply@somewhere.test>
                                To: GPva067Zj sCADSouhPQ4vatf
                                 <recipient@nowhere.test>
                                MIME-Version: 1.0
                                
                                Test message https://www.ingenerator.com/
                                MSG,
            ]
        );
        $this->assertInstanceOf(Email::class, $email);
        $this->assertSame('recipient@nowhere.test', $email->getTo());
        $this->assertSame('inGenerator - Account Activation', $email->getSubject());
        $this->assertSame('Test message https://www.ingenerator.com/', $email->getContent());
        $this->assertSame(['https://www.ingenerator.com/'], $email->getLinks());
    }

    public function test_parses_message()
    {
        $msg = new Swift_Message('About a thing', 'hey, whatever');
        $msg->addTo('some.one@test.com', 'Some One');

        $subject = $this->newSubject();
        $email   = $subject->parseSimulatorCapture(
            ['post' => ['Action' => 'SendRawEmail'], 'raw_data' => $msg->toString()]
        );
        $this->assertSame('some.one@test.com', $email->getTo());
        $this->assertSame('About a thing', $email->getSubject());
        $this->assertSame('hey, whatever', $email->getContent());
    }

    public function test_parses_message_with_html()
    {
        $body = '<html><body><h1 class="heading">Hello</h1><p>Here is a test html message with a <a href="http://example.test">link</a></p></body></html>';
        $msg  = new Swift_Message('Some html', $body, 'text/html');
        $msg->addTo('some.one@test.com', 'Some One');
        $msg->addPart('This,is,a,basic,CSV', 'text/csv');

        $subject = $this->newSubject();
        $email   = $subject->parseSimulatorCapture(
            ['post' => ['Action' => 'SendRawEmail'], 'raw_data' => $msg->toString()]
        );
        $this->assertSame('some.one@test.com', $email->getTo());
        $this->assertSame('Some html', $email->getSubject());
        $this->assertSame(['http://example.test'], $email->getLinks());
    }

    protected function newSubject(): SESSimulatorEmailParser
    {
        return new SESSimulatorEmailParser();
    }

}


src/SES/Transport/SESTransport.php
<?php


namespace Ingenerator\SwiftMailer\SES\Transport;


use Aws\Result;
use Aws\Ses\SesClient;
use Ingenerator\PHPUtils\Monitoring\MetricId;
use Ingenerator\PHPUtils\Monitoring\OperationTimer;
use Ingenerator\SwiftMailer\Common\Transport\AbstractStatelessSwiftTransport;
use Ingenerator\SwiftMailer\Common\Transport\EmailDeliveryFailedException;
use Psr\Log\LoggerInterface;
use Swift_Events_EventDispatcher;
use Swift_Mime_SimpleMessage;
use Throwable;
use function array_keys;
use function get_class;
use function implode;

class SESTransport extends AbstractStatelessSwiftTransport
{
    protected LoggerInterface $logger;

    protected OperationTimer $timer;

    protected SesClient $ses_client;

    public function __construct(
        Swift_Events_EventDispatcher $event_dispatcher,
        SesClient $ses_client,
        OperationTimer $timer,
        LoggerInterface $logger
    ) {
        parent::__construct($event_dispatcher);
        $this->ses_client = $ses_client;
        $this->timer      = $timer;
        $this->logger     = $logger;
    }

    protected function doSend(Swift_Mime_SimpleMessage $message): bool
    {
        return $this->timer->timeOperation(
            function (MetricId $metric) use ($message): bool {
                try {
                    $result = $this->ses_client->sendRawEmail(['RawMessage' => ['Data' => $message->toString()]]);
                    $this->logSuccess($message, $result);

                    return TRUE;
                } catch (Throwable $e) {
                    $metric->setSource('err');
                    $this->logFailure($message, $e);
                    throw new EmailDeliveryFailedException(
                        sprintf('Email send failed: [%s] %s', get_class($e), $e->getMessage()),
                        0,
                        $e
                    );
                }
            },
            'ses-email-sent',
            'ok'
        );
    }

    protected function getRecipientStringForLogging(Swift_Mime_SimpleMessage $message): string
    {
        return implode(',', array_keys($message->getTo()));
    }

    protected function logFailure(Swift_Mime_SimpleMessage $message, Throwable $e): void
    {
        $recipient_string = $this->getRecipientStringForLogging($message);
        $this->logger->error(
            sprintf('Failed to email `%s` - [%s] %s', $recipient_string, get_class($e), $e->getMessage()),
            [
                'exception'      => $e,
                'msg_subject'    => $message->getSubject(),
            ]
        );
    }

    protected function logSuccess(Swift_Mime_SimpleMessage $message, Result $result): void
    {
        $recipient_string = $this->getRecipientStringForLogging($message);
        $this->logger->info(
            sprintf('Emailed %s (%s)', $recipient_string, $message->getSubject()),
            [
                'aws_msg_id'     => $result->get('MessageId'),
            ]
        );
    }

}


src/SES/DevSupport/SESMockEndpoint.php
<?php

namespace Ingenerator\SwiftMailer\SES\DevSupport;

use Ingenerator\PHPUtils\StringEncoding\JSON;

class SESMockEndpoint
{
    public static function endpoint(string $dump_dir): string
    {
        if ( ! is_dir($dump_dir)) {
            mkdir($dump_dir, 0777, TRUE);
        }

        $base_filename = $dump_dir.'/'.uniqid(date('Y-m-d-H-i-s-u'), TRUE);
        if (isset($_POST['RawMessage_Data'])) {
            $raw_data = base64_decode($_POST['RawMessage_Data']);
        } else {
            $raw_data = NULL;
        }

        file_put_contents(
            $base_filename.'.email.json',
            JSON::encode(
                [
                    'headers'  => getallheaders(),
                    'post'     => $_POST,
                    'raw_data' => $raw_data,
                ]
            )
        );

        if ($raw_data) {
            file_put_contents(
                $base_filename.'.email.raw_mime.txt',
                $raw_data
            );
            file_put_contents(
                $base_filename.'.email.quoted_printable.txt',
                quoted_printable_decode($raw_data)
            );
        }

        header('Content-Type: text / xml');
        $message_id = '010201739c605ff5-bfdb0702-12d8-4544-9bc5-'.bin2hex(random_bytes(6)).'-'.bin2hex(random_bytes(3));

        return <<<XML
            <SendRawEmailResponse xmlns="http://ses.amazonaws.com/doc/2010-12-01/">
              <SendRawEmailResult>
                <MessageId>$message_id</MessageId>
              </SendRawEmailResult>
              <ResponseMetadata>
                <RequestId>cbfe3bc1-069e-46d2-9b0d-0c960fc98313</RequestId>
              </ResponseMetadata>
            </SendRawEmailResponse>
            XML;
    }
}


src/SES/Transport/SESTransportFactory.php
<?php
/**
 * @author    Craig Gosman <craig@ingenerator.com>
 * @licence   proprietary
 */


namespace Ingenerator\SwiftMailer\SES\Transport;


use Aws\Ses\SesClient;
use Ingenerator\PHPUtils\Monitoring\MetricsInterface;
use Ingenerator\PHPUtils\Monitoring\OperationTimer;
use Psr\Log\LoggerInterface;
use Swift_DependencyContainer;
use Swift_Events_EventListener;
use function array_merge;

class SESTransportFactory
{
    public static function buildSESClient(array $client_options): SesClient
    {
        return new SesClient(
            array_merge(
                [
                    'region'  => 'eu-west-1',
                    'version' => '2010-12-01',
                    'http'    => [
                        // https://docs.aws.amazon.com/sdk-for-php/v3/developer-guide/guide_configuration.html#config-http
                        // Number of seconds to wait for initial connection
                        'connect_timeout' => 5,
                        // Number of seconds to wait for actual processing
                        'timeout'         => 5,
                    ],
                    'retries' => [
                        // https://docs.aws.amazon.com/sdk-for-php/v3/developer-guide/guide_configuration.html#config-retries
                        // Doesn't retry things that it thinks are unlikely to succeed (e.g. rate limits)
                        // It doesn't seem easy to configure the actual backoff rate or max backoff, we will try 2
                        // requests in quick succession (so could be 20 seconds worst case) before we give up.
                        'mode'         => 'standard',
                        'max_attempts' => 2,
                    ],
                ],
                $client_options
            )
        );
    }

    public static function buildSESTransport(
        SesClient $ses_client,
        OperationTimer $timer,
        LoggerInterface $logger,
        Swift_Events_EventListener ...$plugins
    ): SESTransport {
        $event_dispatcher = Swift_DependencyContainer::getInstance()->lookup('transport.eventdispatcher');
        $transport        = new SESTransport($event_dispatcher, $ses_client, $timer, $logger);

        foreach ($plugins as $plugin) {
            $transport->registerPlugin($plugin);
        }

        return $transport;
    }
}


src/SES/DevSupport/SESSimulatorMailhook.php
<?php

namespace Ingenerator\SwiftMailer\SES\DevSupport;


use Ingenerator\Mailhook\Mailhook;
use Ingenerator\PHPUtils\StringEncoding\JSON;
use RuntimeException;
use function array_filter;
use function basename;
use function copy;
use function file_get_contents;
use function glob;
use function is_dir;
use function mkdir;
use function unlink;

class SESSimulatorMailhook extends Mailhook
{
    protected string $dump_dir;

    public function __construct(string $dump_dir)
    {
        $this->dump_dir = $dump_dir;
        $this->parser   = new SESSimulatorEmailParser;
    }

    public function copyFilesTo(string $dir): void
    {
        if ( ! is_dir($dir)) {
            mkdir($dir, 0777, TRUE);
        }

        foreach ($this->listAllFilesInDumpDir() as $file) {
            copy($file, $dir.'/'.basename($file));
        }
    }

    public function purge(): void
    {
        foreach ($this->listAllFilesInDumpDir() as $file) {
            if ( ! unlink($file)) {
                throw new RuntimeException('Could not delete mail dump at '.$file);
            }
        }
    }

    public function refresh(): void
    {
        $mails     = $this->listAllMailJSONsInDumpDir();
        $new_mails = array_filter($mails, function ($file) { return ! isset($this->emails[$file]); });
        foreach ($new_mails as $file) {
            $this->emails[$file] = $this->parser->parseSimulatorCapture(JSON::decodeArray(file_get_contents($file)));
        }
    }

    protected function listAllFilesInDumpDir(): array
    {
        if ( ! is_dir($this->dump_dir)) {
            return [];
        }

        return glob($this->dump_dir.'/*') ?: [];
    }

    protected function listAllMailJSONsInDumpDir(): array
    {
        if ( ! is_dir($this->dump_dir)) {
            return [];
        }

        return glob($this->dump_dir.'/*.email.json') ?: [];
    }

}


src/Common/Transport/AbstractStatelessSwiftTransport.php
<?php


namespace Ingenerator\SwiftMailer\Common\Transport;


use Ingenerator\PHPUtils\StringEncoding\JSON;
use InvalidArgumentException;
use Swift_Events_EventDispatcher;
use Swift_Events_EventListener;
use Swift_Events_SendEvent;
use Swift_Mime_SimpleMessage;
use Swift_Transport;
use function get_class;

/**
 * Implemented because there seems to be a lot of boilerplate in the Swift_Transport implementations
 * and it's hard to see what's actually the responsibility of the transport in pure "send the message"
 * terms.
 *
 */
abstract class AbstractStatelessSwiftTransport implements Swift_Transport
{
    /**
     * Prop named for parity with Swift's own transports so we can see the duplicate code
     *
     * @var Swift_Events_EventDispatcher
     */
    protected Swift_Events_EventDispatcher $_eventDispatcher;

    public function __construct(Swift_Events_EventDispatcher $event_dispatcher)
    {
        $this->_eventDispatcher = $event_dispatcher;
    }

    public function isStarted(): bool
    {
        return TRUE;
    }

    public function start(): bool
    {
        return TRUE;
    }

    public function stop(): bool
    {
        return TRUE;
    }

    public function ping(): bool
    {
        return TRUE;
    }

    public function send(Swift_Mime_SimpleMessage $message, &$failedRecipients = NULL): int
    {
        $sent_count       = 0;
        $failedRecipients = (array) $failedRecipients;

        if ($evt = $this->_eventDispatcher->createSendEvent($this, $message)) {
            $this->_eventDispatcher->dispatchEvent($evt, 'beforeSendPerformed');
            if ($evt->bubbleCancelled()) {
                return 0;
            }
        }

        // Enforce that the message is kept quite simple as it simplifies our handling
        $this->enforceMaxOneAddress($message->getFrom(), 'from');
        $this->enforceMaxOneAddress($message->getTo(), 'to');

        if ($success = $this->doSend($message)) {
            $sent_count += 1;
        }

        if ($evt) {
            $evt->setResult($success ? Swift_Events_SendEvent::RESULT_SUCCESS : Swift_Events_SendEvent::RESULT_FAILED);
            $this->_eventDispatcher->dispatchEvent($evt, 'sendPerformed');
        }

        return $sent_count;
    }

    /**
     * @param array  $addresses
     * @param string $type
     */
    protected function enforceMaxOneAddress(array $addresses, string $type): void
    {
        if (count($addresses) > 1) {
            throw new InvalidArgumentException(
                get_class($this).' does not support multiple `'.$type.'` address:'.JSON::encode($addresses)
            );
        }
    }

    abstract protected function doSend(Swift_Mime_SimpleMessage $message): bool;

    public function registerPlugin(Swift_Events_EventListener $plugin): void
    {
        $this->_eventDispatcher->bindEventListener($plugin);
    }
}


src/SES/DevSupport/SESSimulatorEmailParser.php
<?php

namespace Ingenerator\SwiftMailer\SES\DevSupport;


use Ingenerator\Mailhook\Email;
use Ingenerator\Mailhook\EmailParser;
use UnexpectedValueException;
use function array_merge;
use function array_pop;
use function preg_match;
use function str_replace;

class SESSimulatorEmailParser extends EmailParser
{
    /**
     * @deprecated Use parseSimulatorCapture
     * @param string $raw_message
     *
     * @return Email
     */
    public function parse($raw_message): Email
    {
        // this method expects $raw_message to be a string of a single message body
        // which doesn't work in this case of an array needing parsed to assert correct action has been called
        throw new \BadMethodCallException('You probably mean to call parseSimulatorCapture');
    }

    public function parseSimulatorCapture(array $simulator_capture): Email
    {
        if ($simulator_capture['post']['Action'] !== 'SendRawEmail') {
            throw new UnexpectedValueException(
                'Expected AWS action SendRawEmail, got '.$simulator_capture['post']['Action']
            );
        }

        $mime_message = $simulator_capture['raw_data'];
        // Seems Swift renders with a windows newline, doesn't trouble SES but bothers mailhook's old parser
        $mime_message = str_replace("\r\n", "\n", $mime_message);

        // Swift also wraps the headers to multiple lines and that also bothers mailhook's very brittle parser
        $mime_message = $this->unWrapHeaderLines($mime_message);

        return parent::parse($mime_message);
    }

    private function unWrapHeaderLines(string $mime_message): string
    {
        /*
         A wrapped header looks like this (pipes added to show the alignment)
            |Date: Thu, 30 Jul 2020 22:55:04 +0100
            |Subject: inGenerator - Account Activation
            |From: inGenerator <noreply@ingenerator.com>
            |To: GPva067Zj sCADSouhPQ4vatf
            | <test+GPva067ZjsCADSouhPQ4vatf.test@ingenerator.com>
            |MIME-Version: 1.0
        */
        $original_lines = explode("\n", $mime_message);
        $fixed_lines    = [];
        while (count($original_lines)) {
            $line = array_shift($original_lines);

            if (trim($line) === '') {
                // Header boundary, stop
                $fixed_lines[] = $line;
                break;
            }

            if (preg_match('/^[^\s]+?: /', $line)) {
                // New header, it's OK
                $fixed_lines[] = $line;
            } else {
                // Wrapped line, merge it up
                $fixed_lines[] = array_pop($fixed_lines).$line;
            }
        }

        $all_lines = array_merge($fixed_lines, $original_lines);

        return implode("\n", $all_lines);
    }

}
